python/pyarrow/src/arrow/python/arrow_to_pandas.cc

```diff
@@ -1074,11 +1074,26 @@ struct ObjectWriterVisitor {
     auto ConvertTimezoneAware = [&](typename Type::c_type value, PyObject** out) {
       PyObject* naive_datetime;
       RETURN_NOT_OK(ConvertTimezoneNaive(value, &naive_datetime));
+
       // convert the timezone naive datetime object to timezone aware
-      *out = PyObject_CallMethod(tzinfo.obj(), "fromutc", "O", naive_datetime);
+      // two step conversion of the datetime mimics Python's code:
+      // dt.replace(tzinfo=datetime.timezone.utc).astimezone(tzinfo)
+      // first step: replacing timezone with timezone.utc (replace method)
+      OwnedRef args(PyTuple_New(0));
+      OwnedRef keywords(PyDict_New());
+      PyDict_SetItemString(keywords.obj(), "tzinfo", PyDateTime_TimeZone_UTC);
+      OwnedRef naive_datetime_replace(PyObject_GetAttrString(naive_datetime, "replace"));
+      OwnedRef datetime_utc(PyObject_Call(naive_datetime_replace.obj(), args.obj(), keywords.obj()));
+      // second step: adjust the datetime to tzinfo timezone (astimezone method)
+      OwnedRef args_astimezone(PyTuple_New(1));
+      PyTuple_SetItem(args_astimezone.obj(), 0, tzinfo.obj());
+      OwnedRef astimezone(PyObject_GetAttrString(datetime_utc.obj(), "astimezone"));
+      *out = PyObject_CallObject(astimezone.obj(), args_astimezone.obj());
```

Comment:
Could this be simplified with [`PyObject_CallMethodOneArg()`](https://docs.python.org/3/c-api/call.html#c.PyObject_CallMethodOneArg)? (since here there are no keyword arguments)

Reply:
Sure, will try 👍 
Plus I am adding a test as we speak (in Python).

Reply:
Ah, for `PyObject_CallMethodOneArg` I need to define `name` as a `PyObject` also, but I get segmentation fault in Python when using:
```cpp
OwnedRef astimezone(Py_BuildValue("astimezone")); 
*out = PyObject_CallMethodOneArg(datetime_utc.obj(), astimezone.obj(), tzinfo.obj());
```

On a separate issue: I also get segmentation fault at the current state of the PR for `test_timestamp_as_object_non_nanosecond()` when `tz` is not None. Am looking into it.

Reply:
`Py_BuildValue` is not meant to create a PyObject string in general like that.

It seems to me that `PyObject_CallMethod(datetime_utc, "astimezone", "O", tzinfo.obj())` (what you had before) should work, though

Reply:
Oh, sorry. OK will try.

Reply:
Thanks for the suggestion! This works and somehow fixes the segmentation issues I had in the tests, locally =)